text/0000-module-resolution.md

```diff
@@ -0,0 +1,458 @@
+        └── button/ (moduleName)
+            └── button.js (moduleEntry)
+```
+
+`DirectoryModuleRecord` must contain `dir` key.
+
+#### NpmModuleRecord
+
+This type of resolution tells the resolver to find an npm package with that name and resolve the modules it might have inside. This process can be recursive (see example at the top).
+
+```json
+{
+    "modules": [
+       { 
+           "npm" : "@ui/components"
+       }     
+    ]
+}
+```
+
+`NpmModuleRecord` must contain `npm` key.
+
+## Module resolution and exposure via NPM packages
+
+So far we have covered how are we going to resolve the modules, however we would want to be able to expose a subset of those modules 
+for other developers in the form of an npm package.
+
+### Exposing modules
+In order to expose a module via npm package, it must be explicity declared in the `expose` array property.
+
+```json
+{
+    "modules": [
+       { 
+           "dir" : "src/modules"
+       }     
+    ],
+    "expose": [
+        "ui-foo",
+        "ui-bar",
+        "ltng-buzz",
+    ]
+}
+```
+::: note
+The reason behind explictly having to expose modules is because every module is a **public api** from the point of view of the consumer, hence by default it should be a restricted list so breaking changes (removal of modules) is intentional and explicit.
+:::
+
+
+### Mapping different modules from different packages
+
+When using multiple npm packages, there can be situations where two or more packages might contain the same module specifier (ex. common-utils,
+so we need to introduce a `map` property to be able to change the specifier so we can disambiguate the behaviour.
+
+```json
+{
+    "modules": [
+       { 
+           "npm" : "lwc-modules-foo",
+           "map": {
+               "common-utils": "foo-common-utils"
+           }
+       },
+       { 
+           "npm" : "lwc-modules-bar",
+           "map": {
+               "common-util:": "bar-common-utils"
+           }
+       }     
+    ]
+}
+```
+
+::: warning
+The module resolver will throw if it detects multiple modules with the same name
+:::
+
+
+## Algorithms
+
+### Preload all modules
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
+  the list of ModuleRecords to be resolved and ModuleRecordEntryList 
+  an empty array.
+  
+  1. Load the LWC configuration file in the rootDir
+    1.1 Search for `lwc.config.js`, if it exist read it.
+    1.2 If `lwc.config.js` is not found search in `package.json`
+    (If both are found the `lwc.config.js` file takes precedence).
+
+  2. Merge all the ModuleRecords found in the config with 
+  the list of initModules. If an alias with the same name 
+  exist pick initModules over the ones resolved in the config.
+
+  3. For each ModuleRecord:
+
+    3.1  Validate that its a valid ModuleRecord type, skip it otherwise.
```

Comment:
If the `ModuleRecord` is invalid we should throw here instead of silently skipping it.